### ISSUE TITLE
Duplicate instance of a static variable when accessed from sub-crate

### ISSUE THREAD

**Issue by Fraser999** [rust-lang/rust#54218]
If I have a `static` variable in a lib and access it from a "sub-crate", the sub-crate appears to create a separate instance of the variable, which is unexpected to me.

I created a minimal project showcasing the issue at https://gitlab.com/Fraser999/subcrate_static_issue:
```
git clone https://gitlab.com/Fraser999/subcrate_static_issue.git
cd subcrate_static_issue
cargo test -- --nocapture
```
The output I'd expect would be:
```
VALUE is 1 after being incremented by subcrate_static_issue.
VALUE is 2 after being incremented by subcrate_static_issue.
VALUE is 3 after being incremented by subcrate.
VALUE is 4 after being incremented by subcrate_static_issue.
VALUE is 5 after being incremented by subcrate.
```
but instead I see:
```
VALUE is 1 after being incremented by subcrate_static_issue.
VALUE is 2 after being incremented by subcrate_static_issue.
VALUE is 1 after being incremented by subcrate.
VALUE is 3 after being incremented by subcrate_static_issue.
VALUE is 2 after being incremented by subcrate.
```
(I'm happy posting the contents of those files directly into this issue if preferred; I just didn't want to clutter the main report too much)

**Comment by jonas-schievink** [comment]
The crate defining the static is being pulled in twice, as the Cargo output or `cargo tree` shows. I'm pretty sure it's expected that statics get duplicated when the defining crate is duplicated.

**Comment by bjorn3** [comment]
This is expected. Rust doesn't allow circular crate dependencies, so cargo treats `subcrate_static_issue` as two separate crates. Once as regular crate at the root and once as dependency of the dev-dependency `subcrate`. If `subcrate` was a regular dependency you would have got an error from cargo. Because cargo compiles `subcrate_static_issue` twice as separate crates, you will get two copies of the static variable.

tl;dr: This is expected behavior of cargo.
